### ISSUE TITLE
test_suite fails on i686

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#789]
This affects the release `4.1.2`
Running  on `i686`
MPI: openmpi (i386, 2.1.1-5)
I could not reproduce this on `x86_64` with the `-m32` flags.

Without patch:
https://kojipkgs.fedoraproject.org//work/tasks/2082/23662082/build.log
With patch #785 
https://kojipkgs.fedoraproject.org//work/tasks/943/23710943/build.log
https://kojipkgs.fedoraproject.org//work/tasks/530/23780530/build.log (currently running)

I am not sure whether this is a race condition - and there was only one failure in the first run, or whether the return code was ignored, and therefore the tests succeeded ...

**Comment by ZedThree** [comment]
This will be a bit tricky to diagnose as I'm not sure many of us have access to anything running a 32-bit OS. I doubt anyone would use one in production either, so this is probably low priority.

That said, the backtraces say it was last in `Mesh::communicate(XZ)`, and they only have `send` and `wait` in common. This could be reasonably serious if either of those have a bug.

**Comment by dschwoerer** [comment]
I currently have access to an 32-bit OS, so I am trying to investigate. I am however still waiting on the test_suite to finish, which has been running for > 32 hours. This however seems to be an issue with openmpi, rather then bout (need to confirm). With mpich the test-suite completes within ~5 min without errors ...

**Comment by dschwoerer** [comment]
I just confirmed. It runs fine on a single thread.

If run on multiple (2,4 or 8) cores - it crashes, or get stuck.

It seems to be stuck here:
One backtrace:
```
#0  0xb410fb80 in opal_timer_linux_get_usec_clock_gettime () from /usr/lib/openmpi/lib/libopen-pal.so.20
#1  0xb40a5a43 in opal_progress () from /usr/lib/openmpi/lib/libopen-pal.so.20
#2  0xb468575d in ompi_request_default_wait_any () from /usr/lib/openmpi/lib/libmpi.so.20
#3  0xb46ce740 in PMPI_Waitany () from /usr/lib/openmpi/lib/libmpi.so.20
#4  0x08114419 in BoutMesh::wait (this=<optimized out>, handle=<optimized out>) at boutmesh.cxx:984
#5  0x0808ca93 in Mesh::communicate (this=0x89b4e70, g=...) at mesh.cxx:182
#6  0x08056c9b in Mesh::communicate<Field3D> (this=0x89b4e70) at ../../../include/bout/mesh.hxx:197
#7  physics_run (t=<optimized out>) at 2fluid.cxx:350
#8  0x0805c0fc in LegacyModel::rhs (this=0x8a785d0, t=1680.3799230726015) at ../../../include/boutmain.hxx:67
#9  0x080d9ebf in PhysicsModel::runRHS (this=0x8a785d0, time=1680.3799230726015) at physicsmodel.cxx:44
#10 0x080d2c94 in Solver::run_rhs (this=<optimized out>, t=<optimized out>) at solver.cxx:1271
#11 0x081205a7 in PvodeSolver::rhs (this=0x8a78970, UNUSED_N=2112, t=1680.3799230726015, udata=0x8bf9e80, dudata=0x8c27360) at pvode.cxx:296
#12 0x0812061b in solver_f (N=2112, t=1680.3799230726015, u=0x8bfe3e0, udot=0x8c06be0, f_data=0x8a78970) at pvode.cxx:332
#13 0x08154654 in pvode::CVSpgmrAtimesDQ (cvode_mem=0x898dad0, v=0x8ae57b0, z=0x8c06be0) at cvspgmr.cpp:462
#14 0x0815527f in pvode::SpgmrSolve (mem=0x8c294f0, A_data=0x898dad0, x=0x8a78570, b=0x8a80d70, pretype=0, gstype=0, delta=1.6850717096511554, max_restarts=0,
    P_data=0x898dad0, sx=0x8a79540, sb=0x8a79540, atimes=0x81545da <pvode::CVSpgmrAtimesDQ(void*, pvode::N_Vector, pvode::N_Vector)>,
    psolve=0x81543f0 <pvode::CVSpgmrPSolve(void*, pvode::N_Vector, pvode::N_Vector, int)>, res_norm=0xbfbc3938, nli=0xbfbc3934, nps=0xbfbc3930) at spgmr.cpp:275
#15 0x08154549 in pvode::CVSpgmrSolve (cv_mem=0x898dad0, b=0x8a80d70, ynow=0x8a807b0, fnow=0x8a79900) at cvspgmr.cpp:390
#16 0x08152978 in pvode::CVNewtonIteration (cv_mem=0x898dad0) at cvode.cpp:1984
#17 pvode::CVnlsNewton (nflag=<optimized out>, cv_mem=0x898dad0) at cvode.cpp:1942
#18 pvode::CVnls (nflag=<optimized out>, cv_mem=0x898dad0) at cvode.cpp:1827
#19 pvode::CVStep (cv_mem=0x898dad0) at cvode.cpp:1358
#20 pvode::CVode (cvode_mem=0x898dad0, tout=1600, yout=0x8a807b0, t=0x8a789c4, itask=0) at cvode.cpp:882
#21 0x0812118c in PvodeSolver::run (this=0x8a78970, tout=1600) at pvode.cxx:244
#22 0x081213ff in PvodeSolver::run (this=<optimized out>) at pvode.cxx:208
#23 0x080d38ac in Solver::solve (this=<optimized out>, NOUT=<optimized out>, TIMESTEP=<optimized out>) at solver.cxx:559
#24 0x08052e5e in main (argc=<optimized out>, argv=<optimized out>) at ../../../include/boutmain.hxx:91
```
and the other one:
```
#0  0xb1401f27 in ompi_coll_libnbc_progress () from /usr/lib/openmpi/lib/openmpi/mca_coll_libnbc.so
#1  0xb411aa8e in opal_progress () from /usr/lib/openmpi/lib/libopen-pal.so.20
#2  0xb46facc5 in ompi_request_default_wait_all () from /usr/lib/openmpi/lib/libmpi.so.20
#3  0xb47603ed in ompi_coll_base_allreduce_intra_recursivedoubling () from /usr/lib/openmpi/lib/libmpi.so.20
#4  0xb142064c in ompi_coll_tuned_allreduce_intra_dec_fixed () from /usr/lib/openmpi/lib/openmpi/mca_coll_tuned.so
#5  0xb470d47e in PMPI_Allreduce () from /usr/lib/openmpi/lib/libmpi.so.20
#6  0x08153d9d in pvode::PVecAllReduce (machEnv=0x88e7d20, op=1, d=<optimized out>) at nvector.cpp:673
#7  pvode::N_VDotProd (x=0x87df900, y=0x87df900) at nvector.cpp:433
#8  0x08155823 in pvode::ModifiedGS (v=0x89197e0, h=0x8924230, k=2, p=10, new_vk_norm=0x88bbd38) at iterativ.cpp:36
#9  0x0815530b in pvode::SpgmrSolve (mem=0x891fba0, A_data=0x8687a10, x=0x88ff3f0, b=0x8773af0, pretype=0, gstype=0, delta=1.6850717096511554, max_restarts=0,
    P_data=0x8687a10, sx=0x8774e80, sb=0x8774e80, atimes=0x81545da <pvode::CVSpgmrAtimesDQ(void*, pvode::N_Vector, pvode::N_Vector)>,
    psolve=0x81543f0 <pvode::CVSpgmrPSolve(void*, pvode::N_Vector, pvode::N_Vector, int)>, res_norm=0xbfe365e8, nli=0xbfe365e4, nps=0xbfe365e0) at spgmr.cpp:302
#10 0x08154549 in pvode::CVSpgmrSolve (cv_mem=0x8687a10, b=0x8773af0, ynow=0x877aae0, fnow=0x877b6a0) at cvspgmr.cpp:390
#11 0x08152978 in pvode::CVNewtonIteration (cv_mem=0x8687a10) at cvode.cpp:1984
#12 pvode::CVnlsNewton (nflag=<optimized out>, cv_mem=0x8687a10) at cvode.cpp:1942
#13 pvode::CVnls (nflag=<optimized out>, cv_mem=0x8687a10) at cvode.cpp:1827
#14 pvode::CVStep (cv_mem=0x8687a10) at cvode.cpp:1358
#15 pvode::CVode (cvode_mem=0x8687a10, tout=1600, yout=0x877aae0, t=0x8773374, itask=0) at cvode.cpp:882
#16 0x0812118c in PvodeSolver::run (this=0x8773320, tout=1600) at pvode.cxx:244
#17 0x081213ff in PvodeSolver::run (this=<optimized out>) at pvode.cxx:208
#18 0x080d38ac in Solver::solve (this=<optimized out>, NOUT=<optimized out>, TIMESTEP=<optimized out>) at solver.cxx:559
#19 0x08052e5e in main (argc=<optimized out>, argv=<optimized out>) at ../../../include/boutmain.hxx:91
```
Not sure this is a bout++ or pvode issue.
Maybe we should switch to a newer pvode?
Further, this is not reproducable, it crashes at different times/not at all. If run on 2 cores for 300 (timestep=10) it crashes in roughly 50% of cases. Could therefore be a race condition.
It took far longer with 8 threads to get stuck.
I will try to get an 32-bit machine next week, which I can make accessible online ...

**Comment by dschwoerer** [comment]
It runs fine on  `openmpi-1.10.3-2.fc24.i686`

**Comment by ZedThree** [comment]
Closing this for now. Will reopen if we get both a reproducer and an i686 system we can actually test on.

I suspect it's something like wrong MPI type for the default `int` on the 32-bit system.

**Comment by dschwoerer** [comment]
Just to follow up on this, the issue seems to be solved by now, so was probably an issue in openmpi.

Current next builds without issues :+1:
